### PR TITLE
Iex h helpers

### PR DESCRIPTION
**Pull request by bbense**
This is a set of patchs for allowing the h command in iex to have a set of back ends for access to documention that is compiled into the existing code. The primary use for this is to provide access Erlang 
module documentation. See: 

https://github.com/elixir-lang/elixir/issues/3589

This patch only provides a stub and largely just duplicates existing functionality using the new backend. There is much work that still needs to be done wrt documentation and whether a Behaviour for the helper modules should be created. If so where does that Behaviour live? The other question is do we provide helpers that are optional. ( The Dash one is awesome... ), or do we simply provide the 
interface and the simple stubs.

### CODE REVIEW

File: lib/iex/lib/iex/h_elixir.ex [hunk 1]
```diff
@@ -0,0 +1,121 @@
+defmodule IEx.H_Elixir do
+	@moduledoc """
```
Comment (lexmag) [inline lib/iex/lib/iex/h_elixir.ex]: Seems the editor made 4-space formatting by mistake (Erlang mode?).

File: lib/iex/test/iex/helpers_test.exs [hunk 2]
```diff
@@ -28,7 +28,8 @@ defmodule IEx.HelpersTest do
            "Could not load module :whatever, got: nofile\n"
 
     assert capture_io(fn -> h :lists end) ==
-           ":lists is an Erlang module and, as such, it does not have Elixir-style docs\n"
+    ":lists is an Erlang module\n and currently there is not a helper installed to provide Erlang documentation\n"
```
Comment (lexmag) [inline lib/iex/test/iex/helpers_test.exs]: It'd be better to keep previous formatting here.

File: lib/iex/lib/iex/introspection.ex [hunk 3]
```diff
@@ -13,18 +13,10 @@ defmodule IEx.Introspection do
...
-            {_, _} ->
-              nodocs(inspect module)
-            _ ->
-              puts_error("#{inspect module} was not compiled with docs")
-          end
-        else
-          puts_error("#{inspect module} is an Erlang module and, as such, it does not have Elixir-style docs")
-        end
+        get_docs(module) |>
+        Enum.map( fn { status, doc_list }  ->
```
Comment (lexmag) [inline lib/iex/lib/iex/introspection.ex]: No spaces after `(`, `{` and before `)`, `}` are needed.

File: lib/iex/lib/iex/config.ex [hunk 4]
```diff
@@ -62,6 +63,7 @@ defmodule IEx.Config do
   defp default_option(:colors), do: [{:enabled, IO.ANSI.enabled?} | default_colors()]
   defp default_option(:inspect), do: []
   defp default_option(:history_size), do: 20
+  defp default_option(:doc_helpers), do: [{:find, :first},{:helpers, [IEx.H_Elixir, IEx.H_Erlang_Stub]}]
```
Comment (lexmag) [inline lib/iex/lib/iex/config.ex]: Helper names looks like an aliens. :bowtie:

File: lib/iex/lib/iex/h_elixir.ex [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+		end 
+	end 
+
+	def get_doc(module, function) when is_atom(module) and is_atom(function) do
+		docs = Code.get_docs(module, :docs)
+		case docs do 
+			nil -> {:not_found, [{"#{inspect module}.#{function}", "No documentation for #{inspect module}.#{function} found\n"}]}
+			_   -> find_doc(docs, module, function)
+		end 
+	end
```
Comment (ericmj) [inline lib/iex/lib/iex/h_elixir.ex]: Change tabs to spaces

File: lib/iex/test/iex/helpers_test.exs [hunk 6]
```diff
@@ -28,7 +28,8 @@ defmodule IEx.HelpersTest do
            "Could not load module :whatever, got: nofile\n"
 
     assert capture_io(fn -> h :lists end) ==
-           ":lists is an Erlang module and, as such, it does not have Elixir-style docs\n"
+           ":lists is an Erlang module\n and currently there is not a helper installed to provide Erlang documentation\n"
```
Comment (ericmj) [inline lib/iex/test/iex/helpers_test.exs]: s/not a/no

Comment (bbense) [inline lib/iex/test/iex/helpers_test.exs]: Should I squash the last 4 commits into one?

Comment (ericmj) [inline lib/iex/test/iex/helpers_test.exs]: Please do

**Comment by bbense** [comment]
I've done some more testing and while this pull request passes all the current tests, it still needs more work to duplicate the current functionality. I'm adding additional tests and working on it. Should I close and resubmit it when it's fully working?

**Comment by josevalim** [comment]
It is fine to leave it open. Just let us know when you are done!

**Comment by bbense** [comment]
I appear to have gotten into some git nightmare mess by attempting to rebase to current master.

I will close this pull request and resubmit with a cleaner commit history.

**Pull request closed without merging**